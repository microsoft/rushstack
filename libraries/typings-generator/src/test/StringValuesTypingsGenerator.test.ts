// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type { StringBufferTerminalProvider, Terminal } from '@rushstack/terminal';
import type {
  IStringValuesTypingsGeneratorBaseOptions,
  IStringValueTypings
} from '../StringValuesTypingsGenerator';

let inputFs: Record<string, string>;
let outputFs: Record<string, string>;

jest.mock('@rushstack/node-core-library', () => {
  const realNcl: typeof import('@rushstack/node-core-library') = jest.requireActual(
    '@rushstack/node-core-library'
  );
  return {
    ...realNcl,
    FileSystem: {
      readFileAsync: async (filePath: string) => {
        const result: string | undefined = inputFs[filePath];
        if (result === undefined) {
          const error: NodeJS.ErrnoException = new Error(
            `Cannot read file ${filePath}`
          ) as NodeJS.ErrnoException;
          error.code = 'ENOENT';
          throw error;
        } else {
          return result;
        }
      },
      writeFileAsync: async (filePath: string, contents: string) => {
        outputFs[filePath] = contents;
      }
    }
  };
});

describe('StringValuesTypingsGenerator', () => {
  beforeEach(() => {
    inputFs = {};
    outputFs = {};
  });

  function runTests(
    baseOptions: IStringValuesTypingsGeneratorBaseOptions,
    extraStringTypings?: Partial<IStringValueTypings>
  ): void {
    it('should generate typings', async () => {
      const [{ StringValuesTypingsGenerator }, { Terminal, StringBufferTerminalProvider }] =
        await Promise.all([import('../StringValuesTypingsGenerator'), import('@rushstack/terminal')]);
      const terminalProvider: StringBufferTerminalProvider = new StringBufferTerminalProvider();
      const terminal: Terminal = new Terminal(terminalProvider);

      inputFs['/src/test.ext'] = '';

      const fileContents: {} = { a: 1 };
      const generator = new StringValuesTypingsGenerator({
        srcFolder: '/src',
        generatedTsFolder: '/out',
        readFile: (filePath: string, relativePath: string) => {
          expect(relativePath).toEqual('test.ext');
          return Promise.resolve(fileContents);
        },
        fileExtensions: ['.ext'],
        parseAndGenerateTypings: (contents: {}, filePath: string, relativePath: string) => {
          expect(contents).toBe(fileContents);
          return {
            typings: [
              {
                exportName: 'test',
                comment: 'test comment\nsecond line'
              }
            ],
            ...extraStringTypings
          };
        },
        terminal,
        ...baseOptions
      });

      await generator.generateTypingsAsync(['test.ext']);
      expect(outputFs).toMatchSnapshot();

      expect(terminalProvider.getOutput()).toEqual('');
      expect(terminalProvider.getWarningOutput()).toEqual('');
      expect(terminalProvider.getErrorOutput()).toEqual('');
      expect(terminalProvider.getVerboseOutput()).toEqual('');
      expect(terminalProvider.getDebugOutput()).toEqual('');
    });
  }

  describe('non-default exports', () => {
    runTests({});
  });

  describe('default exports', () => {
    describe('with { exportAsDefault: true }', () => {
      runTests({ exportAsDefault: true });
    });

    describe("with { exportAsDefault: true, exportAsDefaultInterfaceName: 'IOverride' }", () => {
      runTests({
        exportAsDefault: true,
        exportAsDefaultInterfaceName: 'IOverride'
      });
    });

    describe("with { exportAsDefault: {}, exportAsDefaultInterfaceName: 'IOverride' }", () => {
      runTests({
        exportAsDefault: {},
        exportAsDefaultInterfaceName: 'IOverride'
      });
    });

    describe("with { exportAsDefault: { interfaceName: 'IOverride' }, exportAsDefaultInterfaceName: 'IDeprecated' }", () => {
      runTests({
        exportAsDefault: {
          interfaceName: 'IOverride'
        },
        exportAsDefaultInterfaceName: 'IDeprecated'
      });
    });

    describe("with { exportAsDefault: documentationComment: 'deprecated', interfaceDocumentationComment: 'doc-comment' }", () => {
      runTests({
        exportAsDefault: {
          documentationComment: 'deprecated',
          interfaceDocumentationComment: 'doc-comment'
        }
      });
    });

    describe("with { exportAsDefault: { *DocumentationComment: 'doc-comment\\nsecond line' } }", () => {
      runTests({
        exportAsDefault: {
          interfaceDocumentationComment: 'doc-comment\nsecond line',
          valueDocumentationComment: 'value-comment\nsecond line'
        }
      });
    });

    describe('overrides for individual files', () => {
      describe('with exportAsDefault unset', () => {
        describe('overriding with { exportAsDefault: false }', () => {
          runTests(
            {},
            {
              exportAsDefault: false
            }
          );
        });

        describe("overriding with { interfaceName: 'IOverride' } ", () => {
          runTests(
            {},
            {
              exportAsDefault: {
                interfaceName: 'IOverride'
              }
            }
          );
        });

        describe('overriding with a new doc comment ', () => {
          runTests(
            {},
            {
              exportAsDefault: {
                interfaceDocumentationComment: 'doc-comment\nsecond line',
                valueDocumentationComment: 'value-comment\nsecond line'
              }
            }
          );
        });
      });

      describe('with exportAsDefault set to true', () => {
        describe('overriding with { exportAsDefault: false }', () => {
          runTests(
            {
              exportAsDefault: true
            },
            {
              exportAsDefault: false
            }
          );
        });

        describe("overriding with { interfaceName: 'IOverride' } ", () => {
          runTests(
            {
              exportAsDefault: true
            },
            {
              exportAsDefault: {
                interfaceName: 'IOverride'
              }
            }
          );
        });

        describe('overriding with a new doc comment ', () => {
          runTests(
            {
              exportAsDefault: true
            },
            {
              exportAsDefault: {
                interfaceDocumentationComment: 'doc-comment\nsecond line',
                valueDocumentationComment: 'value-comment\nsecond line'
              }
            }
          );
        });
      });

      describe('with exportAsDefault set to {}', () => {
        describe('overriding with { exportAsDefault: false }', () => {
          runTests(
            {
              exportAsDefault: {}
            },
            {
              exportAsDefault: false
            }
          );
        });

        describe("overriding with { interfaceName: 'IOverride' } ", () => {
          runTests(
            {
              exportAsDefault: {}
            },
            {
              exportAsDefault: {
                interfaceName: 'IOverride'
              }
            }
          );
        });

        describe('overriding with a new doc comment ', () => {
          runTests(
            {
              exportAsDefault: {}
            },
            {
              exportAsDefault: {
                interfaceDocumentationComment: 'doc-comment\nsecond line',
                valueDocumentationComment: 'value-comment\nsecond line'
              }
            }
          );
        });
      });

      describe('with exportAsDefault filled', () => {
        describe('overriding with { exportAsDefault: false }', () => {
          runTests(
            {
              exportAsDefault: {
                interfaceName: 'IBase',
                interfaceDocumentationComment: 'base-comment',
                valueDocumentationComment: 'base-value-comment'
              }
            },
            {
              exportAsDefault: false
            }
          );
        });

        describe("overriding with { interfaceName: 'IOverride' } ", () => {
          runTests(
            {
              exportAsDefault: {
                interfaceName: 'IBase',
                interfaceDocumentationComment: 'base-comment',
                valueDocumentationComment: 'base-value-comment'
              }
            },
            {
              exportAsDefault: {
                interfaceName: 'IOverride'
              }
            }
          );
        });

        describe('overriding with a new doc comment ', () => {
          runTests(
            {
              exportAsDefault: {
                interfaceName: 'IBase',
                interfaceDocumentationComment: 'base-comment',
                valueDocumentationComment: 'base-value-comment'
              }
            },
            {
              exportAsDefault: {
                interfaceDocumentationComment: 'doc-comment\nsecond line',
                valueDocumentationComment: 'value-comment\nsecond line'
              }
            }
          );
        });
      });
    });
  });
});
