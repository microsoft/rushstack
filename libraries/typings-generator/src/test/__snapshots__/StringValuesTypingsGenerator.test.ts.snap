// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StringValuesTypingsGenerator default exports overrides for individual files with exportAsDefault filled overriding with { exportAsDefault: false } should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

/**
 * test comment
second line
 */
export declare const test: string;
",
}
`;

exports[`StringValuesTypingsGenerator default exports overrides for individual files with exportAsDefault filled overriding with { interfaceName: 'IOverride' }  should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

/**
 * base-comment
 */
export interface IOverride {
  /**
   * test comment
second line
   */
  'test': string;

}

/**
 * base-value-comment
 */
declare const strings: IOverride;

export default strings;",
}
`;

exports[`StringValuesTypingsGenerator default exports overrides for individual files with exportAsDefault filled overriding with a new doc comment  should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

/**
 * doc-comment
 * second line
 */
export interface IBase {
  /**
   * test comment
second line
   */
  'test': string;

}

/**
 * value-comment
 * second line
 */
declare const strings: IBase;

export default strings;",
}
`;

exports[`StringValuesTypingsGenerator default exports overrides for individual files with exportAsDefault set to {} overriding with { exportAsDefault: false } should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

/**
 * test comment
second line
 */
export declare const test: string;
",
}
`;

exports[`StringValuesTypingsGenerator default exports overrides for individual files with exportAsDefault set to {} overriding with { interfaceName: 'IOverride' }  should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface IOverride {
  /**
   * test comment
second line
   */
  'test': string;

}

declare const strings: IOverride;

export default strings;",
}
`;

exports[`StringValuesTypingsGenerator default exports overrides for individual files with exportAsDefault set to {} overriding with a new doc comment  should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

/**
 * doc-comment
 * second line
 */
export interface IExport {
  /**
   * test comment
second line
   */
  'test': string;

}

/**
 * value-comment
 * second line
 */
declare const strings: IExport;

export default strings;",
}
`;

exports[`StringValuesTypingsGenerator default exports overrides for individual files with exportAsDefault set to true overriding with { exportAsDefault: false } should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

/**
 * test comment
second line
 */
export declare const test: string;
",
}
`;

exports[`StringValuesTypingsGenerator default exports overrides for individual files with exportAsDefault set to true overriding with { interfaceName: 'IOverride' }  should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface IOverride {
  /**
   * test comment
second line
   */
  'test': string;

}

declare const strings: IOverride;

export default strings;",
}
`;

exports[`StringValuesTypingsGenerator default exports overrides for individual files with exportAsDefault set to true overriding with a new doc comment  should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

/**
 * doc-comment
 * second line
 */
export interface IExport {
  /**
   * test comment
second line
   */
  'test': string;

}

/**
 * value-comment
 * second line
 */
declare const strings: IExport;

export default strings;",
}
`;

exports[`StringValuesTypingsGenerator default exports overrides for individual files with exportAsDefault unset overriding with { exportAsDefault: false } should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

/**
 * test comment
second line
 */
export declare const test: string;
",
}
`;

exports[`StringValuesTypingsGenerator default exports overrides for individual files with exportAsDefault unset overriding with { interfaceName: 'IOverride' }  should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface IOverride {
  /**
   * test comment
second line
   */
  'test': string;

}

declare const strings: IOverride;

export default strings;",
}
`;

exports[`StringValuesTypingsGenerator default exports overrides for individual files with exportAsDefault unset overriding with a new doc comment  should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

/**
 * doc-comment
 * second line
 */
export interface IExport {
  /**
   * test comment
second line
   */
  'test': string;

}

/**
 * value-comment
 * second line
 */
declare const strings: IExport;

export default strings;",
}
`;

exports[`StringValuesTypingsGenerator default exports with { exportAsDefault: { *DocumentationComment: 'doc-comment\\nsecond line' } } should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

/**
 * doc-comment
 * second line
 */
export interface IExport {
  /**
   * test comment
second line
   */
  'test': string;

}

/**
 * value-comment
 * second line
 */
declare const strings: IExport;

export default strings;",
}
`;

exports[`StringValuesTypingsGenerator default exports with { exportAsDefault: { interfaceName: 'IOverride' }, exportAsDefaultInterfaceName: 'IDeprecated' } should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface IOverride {
  /**
   * test comment
second line
   */
  'test': string;

}

declare const strings: IOverride;

export default strings;",
}
`;

exports[`StringValuesTypingsGenerator default exports with { exportAsDefault: {}, exportAsDefaultInterfaceName: 'IOverride' } should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface IOverride {
  /**
   * test comment
second line
   */
  'test': string;

}

declare const strings: IOverride;

export default strings;",
}
`;

exports[`StringValuesTypingsGenerator default exports with { exportAsDefault: documentationComment: 'deprecated', interfaceDocumentationComment: 'doc-comment' } should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

/**
 * doc-comment
 */
export interface IExport {
  /**
   * test comment
second line
   */
  'test': string;

}

declare const strings: IExport;

export default strings;",
}
`;

exports[`StringValuesTypingsGenerator default exports with { exportAsDefault: true } should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface IExport {
  /**
   * test comment
second line
   */
  'test': string;

}

declare const strings: IExport;

export default strings;",
}
`;

exports[`StringValuesTypingsGenerator default exports with { exportAsDefault: true, exportAsDefaultInterfaceName: 'IOverride' } should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

export interface IOverride {
  /**
   * test comment
second line
   */
  'test': string;

}

declare const strings: IOverride;

export default strings;",
}
`;

exports[`StringValuesTypingsGenerator non-default exports should generate typings 1`] = `
Object {
  "/out/test.ext.d.ts": "// This file was generated by a tool. Modifying it will produce unexpected behavior

/**
 * test comment
second line
 */
export declare const test: string;
",
}
`;
